### ISSUE TITLE
Eshell fails to complete paths with `ls`

### ISSUE THREAD

**Issue by paulapatience** [hlissner/doom-emacs#5125]
### What did you expect to happen?

When typing `ls` followed by space and then tab, a list of paths available for completion should appear.

### What actually happened?

Nothing happens, i.e., neither error messages nor completion candidates appear on the screen.

### Describe your attempts to resolve the issue

Manually running `(eshell-cmpl-initialize)` in an Eshell buffer fixes the problem. However, I am unable to add this function to `eshell-mode-hook`. From what I can tell, `eshell-mode-hook` does not get added to properly. (None of the `add-hook`s in `modules/term/eshell/config.el` seem to be effective in Eshell buffers.)

### Steps to reproduce

1. Open Emacs.
2. Open Eshell with `SPC o E`.
3. Type `ls`, space, then tab.
4. Nothing will complete.

### System Information

https://pastebin.com/54QHCL90

**Comment by hlissner** [comment]
I cannot reproduce this, so there is likely an error occurring at startup that is preventing the eshell module from initializing fully. Check your \*Messages\* log buffer for suspicious warnings or errors. `SPC h e` or `C-h e` will open this buffer.

**Comment by paulapatience** [comment]
I tried again but this time with vanilla Doom (running `git clean -dxf` in `~/.emacs.d` and then `doom install`) with the default `init.el` and `config.el` and with Emacs 27.2 and 28.0.50 and still get no completions. `*Messages*` contains no warnings or errors. I'm quite stumped. What's curious is that `cd` does offer directory completions, and command names get completed as well, though I suppose that's just because the default `eshell-mode-hook` allows it.

It may be useful to note that I've installed Emacs through Guix, but that was also the case when I didn't have this problem a few months ago.

**Comment by paulapatience** [comment]
What's curious is that the value of `eshell-mode-hook` in a non-Eshell buffer, for example `modules/term/eshell/config.el`, contains all the additions of the `(after! eshell ...)` block:

```
(#[nil "\301\300!\210\302\211\207"
      [eldoc-documentation-function make-local-variable esh-help-eldoc-command]
      2]
 hide-mode-line-mode
 +eshell-remove-fringes-h
 +eshell-enable-text-wrapping-h
 smartparens-mode
 +eshell-init-h
 doom-mark-buffer-as-real-h
 evil-collection-eshell-next-prompt-on-insert
 fish-completion-mode
 eshell-syntax-highlighting-mode
 doom--setq-hscroll-margin-for-eshell-mode-h)
```

But in an Eshell buffer, `eshell-mode-hook` is changed:

```
(#<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_26>
 t)
```

**Comment by hlissner** [comment]
> But in an Eshell buffer, `eshell-mode-hook` is changed:

That's because an anonymous hook was added to it buffer-locally. The `t` means "run the global hooks here after the local hooks are done".

**Comment by paulapatience** [comment]
Thank you for that information — I hadn't thought to read up on hooks — which allowed me to do some more debugging and determine that the hooks actually do get run. The reason I thought they didn't is because when adding `eshell-cmpl-initialize` to the beginning of `eshell-mode-hook`, it has no effect; it takes effect only when it's at the end (I didn't test when it's in the middle). Also, it seems that appending it with `add-hook!`'s `:append` actually adds it to the beginning the first time — see the code below.

For reference, the following code successfully adds `eshell-cmpl-initialize` to `eshell-mode-hook` and results in (my) Eshell completing paths with `ls`:

```emacs-lisp
(after! eshell
  ;; FIXME: Path-completion, for example with "ls", is disabled until
  ;; `eshell-cmpl-initialize' is called.
  (add-hook! eshell-mode :append #'eshell-cmpl-initialize)
  ;; For some reason, the first `add-hook!' adds
  ;; `eshell-cmpl-initialize' to the beginning, even with `:append'.
  ;; Remove it and add it again to truly append it.
  ;; (When it is at the beginning, it fails to enable completions.)
  (remove-hook! eshell-mode #'eshell-cmpl-initialize)
  (add-hook! eshell-mode :append #'eshell-cmpl-initialize)

  ;; Make sure the hooks were run.
  (add-hook! eshell-mode :append
    (defun my-post-eshell-mode-hook-h ()
      (message "Ran hooks in `eshell-mode-hook'."))))
```

The question that remains is why I need to do this to get path completion. Also, I don't know if calling `eshell-cmpl-initialize` is the way Eshell completion is supposed to be enabled in Doom Emacs or if there is another way (e.g., for Fish completion, Company completion, etc.).

**Comment by hlissner** [comment]
Whoops, linked the wrong issue.

**Comment by paulapatience** [comment]
After upgrading Doom recently (maybe after the first Vertico merge?), I looked into this again and determined that what was causing problems was having company-indent-or-complete-common bound to [tab] (probably, anyway — I didn't perform any kind of structured testing). However, I mentioned above that I had tested with a vanilla Doom Emacs and default init.el and config.el files — perhaps my other configuration files were getting loaded anyway (i.e., including the one where I rebound [tab])? Or perhaps there actually was an issue that got magically fixed in the meantime. Regardless, I'm closing this issue because it is not an issue for me anymore.
